src/main/java/com/google/sps/servlets/RoutesStoring.java

```
@@ -0,0 +1,107 @@
+    // Check if user is logged in.
+    if (!userService.isUserLoggedIn()) {
+      response.setContentType("text/html;");
+      response.getWriter().println("<p>ERROR: You are not logged in</p>");
+      response.getWriter().println("<a href=\"index.html\">Go to home page</button>");
+      return;
+    }
+
+    // Read request body.
+    String requestBody =
+        request.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
+
+    // Convert it to Gson Object.
+    Gson gson = new Gson();
+    RouteData gsonObject = gson.fromJson(requestBody, RouteData.class);
+    Marker[] routeMarkers = gsonObject.getMarkersData();
+    String routeName = gsonObject.getRouteName();
+
+    Key userKey = KeyFactory.createKey("User", userService.getCurrentUser().getUserId());
+
+    try {
+      Entity userEntity = datastore.get(userKey);
+      // Create new route entity and make it child of the user.
+      Entity routeEntity = new Entity("Route", userEntity.getKey());
+      routeEntity.setProperty("name", routeName);
+      datastore.put(routeEntity);
+
+      for (Marker marker : routeMarkers) {
+        // Create entities for markers and make them children of the route.
+        Entity markerEntity = new Entity("Marker", routeEntity.getKey());
+
+        markerEntity.setProperty("lat", marker.getLat());
+        markerEntity.setProperty("lng", marker.getLng());
+        markerEntity.setProperty("visitHour", marker.getVisitHour());
+        markerEntity.setProperty("visitMinute", marker.getVisitMinute());
+        markerEntity.setProperty("leaveHour", marker.getLeaveHour());
+        markerEntity.setProperty("leaveMinute", marker.getLeaveMinute());
+        markerEntity.setProperty("markerName", marker.getMarkerName());
+
+        // Store new marker.
+        datastore.put(markerEntity);
+      }
+      response.sendRedirect("index.html");
+    } catch (Exception e) {
+      response.setContentType("text/html;");
```

Should rather return an error status message in JSON which can be handled by the javascript that executed the fetch. As mentioned in other comments, this can be done in a follow-up PR.

---

Done